app/assets/javascripts/index/directions.js

```diff
@@ -96,15 +91,36 @@ OSM.Directions = function (map) {
     return h + ":" + (m < 10 ? "0" : "") + m;
   }
 
-  function findEngine(id) {
-    return engines.findIndex(function (engine) {
-      return engine.id === id;
-    });
-  }
-
-  function setEngine(index) {
-    chosenEngine = engines[index];
-    select.val(index);
+  function setEngine(id) {
+    const engines = OSM.Directions.engines;
+    const desired = engines.find(engine => engine.id() === id);
+    if (!desired || (chosenEngine && chosenEngine.id() === id)) return;
+    chosenEngine = desired;
+
+    const modes = engines
+      .filter(engine => engine.provider === chosenEngine.provider)
+      .map(engine => engine.mode)
+      .sort((a, b) => I18n.t("javascripts.directions.modes." + a).localeCompare(I18n.t("javascripts.directions.modes." + b)));
+    modeGroup.html("");
+    for (const mode of new Set(modes)) {
+      modeGroup.append(`<input type="radio" class="btn-check" name="modes" id="${mode}" autocomplete="off">`);
+      modeGroup.append(`<label class="btn btn-outline-secondary px-2" for="${mode}" title="${
+        I18n.t("javascripts.directions.modes." + mode)
+      }"><svg class="d-block" width="16" height="16" fill="currentColor"><path d="${modeIconPaths[mode]}"></path></svg></label>`);
+    }
+    $(".routing_modes input#" + chosenEngine.mode).prop("checked", true);
+
+    const providers = engines
+      .filter(engine => engine.mode === chosenEngine.mode)
+      .map(engine => engine.provider)
+      .sort((a, b) => I18n.t("javascripts.directions.providers." + a).localeCompare(I18n.t("javascripts.directions.providers." + b)));
+    select.html("");
+    for (const provider of new Set(providers)) {
+      select.append(`<option value="${provider}">${
+        I18n.t("javascripts.directions.providers." + provider)
```

Comment:
I know this *let's interpolate unescaped html from external source* was in the original code, but you don't have to replicate it.